src/main/java/com/microsoft/sqlserver/jdbc/NTLMAuthentication.java

```diff
@@ -0,0 +1,906 @@
+     * Section 2.2.2.7 NTLM v2: NTLMv2_CLIENT_CHALLENGE:
+     *
+     * clientChallenge blob consists of:
+     *     Responserversion,
+     *     HiResponserversion,    NTLM_CLIENT_CHALLENGE_RESPONSE_TYPE,
+     *     Z(6),                  NTLM_CLIENT_CHALLENGE_RESERVED1, NTLM_CLIENT_CHALLENGE_RESERVED2
+     *     Time,                  current timestamp
+     *     ClientChallenge,       clientNonce
+     *     Z(4),                  NTLM_CLIENT_CHALLENGE_RESERVED2
+     *     ServerName,            CHALLENGE_MESSAGE.TargetInfo (from 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server)
+     *     Z(4))                  EOL
+     * </pre>
+     *
+     * @param clientNonce
+     *        client challenge nonce
+     * @return client challenge blob
+     */
+    private byte[] generateClientChallengeBlob(final byte[] clientNonce) {
+
+        // timestamp is number of 100 nanosecond ticks since Windows Epoch
+        ByteBuffer time = ByteBuffer.allocate(8).order(ByteOrder.LITTLE_ENDIAN);
+        time.putLong((TimeUnit.SECONDS.toNanos(Instant.now().getEpochSecond() + WINDOWS_EPOCH_DIFF)) / 100);
+        byte[] currentTime = time.array();
+
+        context.token = ByteBuffer
+                .allocate(NTLM_CLIENT_CHALLENGE_RESPONSE_TYPE.length + NTLM_CLIENT_CHALLENGE_RESERVED1.length
+                        + NTLM_CLIENT_CHALLENGE_RESERVED2.length + currentTime.length + NTLM_CLIENT_NONCE_LENGTH
+                        + NTLM_CLIENT_CHALLENGE_RESERVED3.length + context.targetInfo.length + NTLM_MIC_AVP_LENGTH)
+                .order(ByteOrder.LITTLE_ENDIAN);
+
+        context.token.put(NTLM_CLIENT_CHALLENGE_RESPONSE_TYPE);
+        context.token.put(NTLM_CLIENT_CHALLENGE_RESERVED1);
+        context.token.put(NTLM_CLIENT_CHALLENGE_RESERVED2);
+
+        context.token.put(currentTime, 0, NTLM_TIMESTAMP_LENGTH);
+        context.token.put(clientNonce, 0, NTLM_CLIENT_NONCE_LENGTH);
+        context.token.put(NTLM_CLIENT_CHALLENGE_RESERVED3);
+
+        /**
+         * Section 3.1.5.1.2 Client Receives a CHALLENGE_MESSAGE from the Server
+         *
+         * If the CHALLENGE_MESSAGE TargetInfo field has an MsvAvTimestamp present, the client SHOULD provide a MIC
+         */
+        if (null == context.timestamp || 0 >= context.timestamp.length) {
+            context.token.put(context.targetInfo, 0, context.targetInfo.length);
+        } else {
+            ByteBuffer newTargetInfo = ByteBuffer.allocate(context.targetInfo.length + NTLM_MIC_AVP_LENGTH)
+                    .order(ByteOrder.LITTLE_ENDIAN);
+
+            // copy targetInfo up to NTLM_AVID_MSVAVEOL
+            newTargetInfo.put(context.targetInfo, 0, context.targetInfo.length - 4);
+
+            // MIC flag
+            newTargetInfo.putShort(NTLM_AVID_MSVAVFLAGS);
+            newTargetInfo.putShort((short) 4);
+            newTargetInfo.putInt((int) NTLM_AVID_VALUE_MIC);
+
+            // EOL
+            newTargetInfo.putShort(NTLM_AVID_MSVAVEOL);

```

Comment:
I don't think these two lines are required, you are writing zeroes when the array is already initialized to zeroes.

Reply:
it makes it more clear, you just happen to know EOL is 0 but it may not be :)